### ISSUE TITLE
Port twisted.internet.serialport to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8099]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)| @thijstriemstra reported|
|-|-|
|Trac ID|trac#8099|
|Type|task|
|Created|2015-11-07 17:31:13Z|
|Branch|https://github.com/twisted/twisted/tree/serialport-py3-8099|

Was playing with the txgsm project on Python 3.4, which uses `twisted.internet.serialport`, and noticed it didn't install `twisted/internet/serialport.py` and `twisted/internet/_posixserialport.py` in my virtualenv.

This [stackoverflow post](http://stackoverflow.com/questions/24291443/why-isnt-serialport-py-installed-by-default) describes the same problem, but the solution is simple, port it to python 3 and include it in `twisted.python.dist3` so it gets shipped.

Attachments:

* [8099.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d42/d42a52fe0df9e0f7c005e0ff2ceb7db3010a6d3d/b1f2d2abb1f9e75f7d423a53500816445759bbd2.patch) (2761 bytes) - added by thijs on 2015-11-07 18:11:43Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__8099 8099
type__task task
reporter__thijs thijs
priority__normal normal
milestone__Python_3_x Python-3.x
branch__branches_serialport_py3_8099 branches/serialport-py3-8099
branch_author__adiroiban adiroiban
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1446917473280240 1446917473280240
changetime__1447663348921932 1447663348921932
version__None None
owner__thijs thijs
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_1"></a>|@thijstriemstra commented|
|-|-|

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_2"></a>|@adiroiban commented|
|-|-|

(In [46115]) Branching to serialport-py3-8099.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban set owner to @thijstriemstra|
|-|-|


Many thanks for the patch!

The patch also needs a news file fragment so that we have this new feature mentioned in the release notes.


I have also added the win32 serial code to the list of ported code so that it will be installed on windows.

I have sent the patch to buildbot.

If all is ok, will merge.

I have added the news fragment. Please check that it make sense.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_4"></a>|@adiroiban commented|
|-|-|

While tests are green (most of them) there are a few problems with twisted/internet/test/test_serialport.py

The tests are pretty thin and on top of that they are not executed on our python 3 builders.

Twisted is not yet ported on Windows with Python3.

I will check with the other Twisted developers and see if we can get pyserial installed/updated our the testing infrastructure.

I would like to have the test run on Python3 before merging this :)

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_5"></a>|@thijstriemstra set owner to @adiroiban|
|-|-|

Thanks for the update adiroiban. I realize there's not a lot of tests but everything seems to work fine on posix. Any news on the pyserial installations? I suppose I'll assign it back to you.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_6"></a>|@adiroiban set owner to @thijstriemstra|
|-|-|

I have installed pyserial on our buildslaves and the single test looks good so far.

Will merge once all tests are done.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_7"></a>|@adiroiban set status to `closed`|
|-|-|

(In [46227]) Merge serialport-py3-8099: twisted.internet.serialport port on Python 3.

Author: thijs
Reviewer: adiroiban
Fixes: [#8099](https://github.com/twisted/twisted/issues/8099)
